src/erlazure_ad_auth.erl

```diff
@@ -0,0 +1,93 @@
+-module(erlazure_ad_auth).
+
+%% this module implements Azure Active Directory Authentication Protocols
+%% which are described at https://msdn.microsoft.com/en-us/library/azure/dn151124.aspx
+
+%% there is design decision to use Options proplist
+%% for all the not required parameters
+%% in order to shorten amount of functions with different number of arguments
+
+
+%% This client is common to all tenants
+%% This is used for username password logins
+-define(CLIENT_ID, "04b07795-8ddb-461a-bbee-02f9e1bf7b46").
+%% default resource
+-define(RESOURCE, "https://management.core.windows.net/").
+%% default authority
+-define(AUTHORITY, "login.microsoftonline.com").
+
+-export([
+    oauth2_acquire_token_with_username_password/4,
+    oauth2_acquire_token_with_client_credentials/4
+]).
+
+-include("erlazure.hrl").
+
+%% interface design questions:
+%% support Config in erlcloud style to be able to change endpoints?
+%%     also with possible retry mechanism
+
+-spec oauth2_acquire_token_with_username_password(
+    string(), string(), string(), proplists:proplist()) ->
+    {ok, ad_oauth2_token()} | {error, Reason :: term()}.
+oauth2_acquire_token_with_username_password(ADId, Username, Password, Options) ->
+    Authority = proplists:get_value(authority, Options, ?AUTHORITY),
+    Params = [
+        {"grant_type", "password"},
+        {"client_id", proplists:get_value(client_id, Options, ?CLIENT_ID)},
+        {"resource", proplists:get_value(resource, Options, ?RESOURCE)},
+        {"username", Username},
+        {"password", Password}
+    ],
+    request(Authority, ADId, Params).
+
+-spec oauth2_acquire_token_with_client_credentials(
+    string(), string(), string(), proplists:proplist()) ->
+    {ok, ad_oauth2_token()} | {error, Reason :: term()}.
+oauth2_acquire_token_with_client_credentials(ADId, ClientId, ClientSecret, Options) ->
+    Authority = proplists:get_value(authority, Options, ?AUTHORITY),
+    Params = [
+        {"grant_type", "client_credentials"},
+        {"resource", proplists:get_value(resource, Options, ?RESOURCE)},
+        {"client_id", ClientId},
+        {"client_secret", ClientSecret}
+    ],
+    request(Authority, ADId, Params).
+
+
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+
+request(Authority, ADId, Params) ->
```

Comment:
as @kkuzmin pointed out, we can separate this into some other module in order to be able implement retry/throttling logic in the same place